src/lib/reasoners/fun_sat.ml

```diff
@@ -1654,14 +1642,29 @@ module Make (Th : Theory.S) = struct
               (mk_gf (E.neg guard_to_neg) "" true true,Ex.empty)
               acc.guards.guards
           in
+          let gamma =
+            (* If we made a check-sat and we want to work again on the
+               environment, we have to remove the guard from gamma. *)
+            match ME.find_opt guard_to_neg acc.gamma with
+            | None -> acc.gamma
+            | Some (_, _, dlvl, plvl) ->
+              ME.filter
+                (fun _ (_, _, dlvl', plvl') ->
+                   not (dlvl = dlvl' && plvl' >= plvl))
```

Comment:
Shouldn't we also remove things at later decision levels? I'd have expected `not (dlvl' > dlvl || (dlvl' = dlvl && plvl' >= plvp))`.

Reply:
My comment above still stands. Is the simpler test enough because we pop levels one by one?